test/gleam/dynamic_test.gleam

```diff
@@ -744,6 +744,20 @@ pub fn map_test() {
   |> dynamic.from
   |> dynamic.map(dynamic.string, dynamic.int)
   |> should.equal(Error([DecodeError(expected: "Map", found: "Int", path: [])]))
+
+  #()
```

Comment:
Note can't write the same test for a list literal. because in JS they are represented as Objects.

I think this is still a fair implementation. 
1. The found field in the error is a string for human debugging.
2. The type being decoded should be assumed to be JS. 

The result is if in Javascript a gleam function is called with `[]` as a dynamic value the error message will call this a tuple and not a List. This test is at least consistent with that behaviour

Reply:
I think we're going to need to check to ensure that these values are not Gleam data types prior to attempting to decode as a Map or JS object. If a list can be decoded as an object then a decoder which attempts to decode a map or a list would never be able to get to the list case.

Reply:
Or perhaps it should only work for objects without a custom prototype?

Reply:
That's a good solution